### PR TITLE
Helpers for filter counts

### PR DESCRIPTION
**Pull request by jamesckemp**
I've done 2 things here (as per our conversations earlier today):
1. Made it use the post_types set in the main query.
2. Added a filter to the "where" clause. This means I can hook in and add and additional parameter to the where clause to exclude specific posts.

Doing the above has fixed any filter count issues I have, seemingly. I'd really appreciate it if we can do this (or similar).

### CODE REVIEW

File: includes/widgets/class-wc-widget-layered-nav.php [hunk 1]
```diff
@@ -352,12 +352,14 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
 		$tax_query      = new WP_Tax_Query( $tax_query );
 		$meta_query_sql = $meta_query->get_sql( 'post', $wpdb->posts, 'ID' );
 		$tax_query_sql  = $tax_query->get_sql( $wpdb->posts, 'ID' );
+		$post_type      = (array)$wp_the_query->query_vars['post_type'];
```
Comment (mikejolley) [inline includes/widgets/class-wc-widget-layered-nav.php]: Please handle arrays vs string here

File: includes/widgets/class-wc-widget-layered-nav.php [hunk 2]
```diff
@@ -352,12 +352,14 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
...
+		$where          = "WHERE {$wpdb->posts}.post_type IN ('" . implode("','", $post_type) . "') AND {$wpdb->posts}.post_status = 'publish'";
```
Comment (mikejolley) [inline includes/widgets/class-wc-widget-layered-nav.php]: needs array_map esc_sql() on the post types

File: includes/widgets/class-wc-widget-layered-nav.php [hunk 3]
```diff
@@ -352,12 +352,14 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
...
 
 		$sql  = "
 			SELECT COUNT( {$wpdb->posts}.ID ) as term_count, term_count_relationships.term_taxonomy_id as term_count_id FROM {$wpdb->posts}
 			INNER JOIN {$wpdb->term_relationships} AS term_count_relationships ON ({$wpdb->posts}.ID = term_count_relationships.object_id)
 			" . $tax_query_sql['join'] . $meta_query_sql['join'] . "
-			WHERE {$wpdb->posts}.post_type = 'product' AND {$wpdb->posts}.post_status = 'publish'
+			" . apply_filters('woocommerce_filtered_term_product_counts_where_clause', $where ) . "
```
Comment (mikejolley) [inline includes/widgets/class-wc-widget-layered-nav.php]: Would be cleaner to have this above during $where generation.

**Comment by mikejolley** [comment]
@jamesckemp see feedback

**Comment by jamesckemp** [comment]
@mikejolley Thanks, updated. Not sure if the filter name I've used is too long? Seems quite a mouthful!

File: includes/widgets/class-wc-widget-layered-nav.php [hunk 4]
```diff
@@ -352,12 +352,14 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
 		$tax_query      = new WP_Tax_Query( $tax_query );
 		$meta_query_sql = $meta_query->get_sql( 'post', $wpdb->posts, 'ID' );
 		$tax_query_sql  = $tax_query->get_sql( $wpdb->posts, 'ID' );
+		$post_type      = !is_array( $wp_the_query->query_vars['post_type'] ) ? array( $wp_the_query->query_vars['post_type'] ) : $wp_the_query->query_vars['post_type'];
```
Comment (mikejolley) [inline includes/widgets/class-wc-widget-layered-nav.php]: reverse the logic here for readability

File: includes/widgets/class-wc-widget-layered-nav.php [hunk 5]
```diff
@@ -352,12 +352,14 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
...
+		$where          = apply_filters('woocommerce_filtered_term_product_counts_where_clause', "WHERE {$wpdb->posts}.post_type IN ('" . implode("','", array_map( 'esc_sql', $post_type )) . "') AND {$wpdb->posts}.post_status = 'publish'");
```
Comment (mikejolley) [inline includes/widgets/class-wc-widget-layered-nav.php]: Use woocommerce_get_filtered_term_product_counts_query_where, and pass $term_ids taxonomy and query type to it as arguments.

Brackets should also have a single space around them.

**Comment by jamesckemp** [comment]
@mikejolley Feedback implemented

**Comment by mikejolley** [comment]
@jamesckemp might want to re-test that, I found a bug in the logic regarding term_ids.

**Comment by jamesckemp** [comment]
@mikejolley All seems OK to me - returning the correct results with my plugin enabled/disabled. What were you seeing?

**Comment by mikejolley** [comment]
Just something I noticed on a site which hasn't had term splitting kick in yet. Did not know if my commits would break yours.

**Comment by jamesckemp** [comment]
Ah, I see. Let me double check.

**Comment by jamesckemp** [comment]
@mikejolley Yep, still works as expected - your latest updates have fixed an issue I spotted when using the price filter at the same time (count was wrong). So all looking good!

Did you want me to submit a new PR with the changes, or is there an easy way to merge your new master into my modified file? (I've done all the changes on GitHub directly, so no local fork.)

**Comment by claudiosmweb** [comment]
The post type here should be always be for `product`, if not you can create a new widget.

**Comment by mikejolley** [comment]
@claudiosmweb the use case is to enabling filtering on product variations too.

**Comment by claudiosmweb** [comment]
@mikejolley so we should allow only `product` and `product_variation`.

**Comment by jamesckemp** [comment]
@claudiosmweb @mikejolley That's fine for my scenario, but is there any other scenario where someone might want something different? Would there be any need to limit it explicitly?

**Comment by mikejolley** [comment]
For some reason this PR overwrites my fixes. I'll have to apply manually. The post type code can be removed  - use the filter to add variations.

**Comment by jamesckemp** [comment]
@mikejolley Yeah, that'd be fine actually.

**Pull request closed without merging**